taichi/system/memory_pool.cpp

```diff
@@ -1,10 +1,12 @@
 #include "memory_pool.h"
 #include "taichi/system/timer.h"
 #include "taichi/backends/cuda/cuda_driver.h"
+#include "taichi/backends/cuda/cuda_device.h"
 
 TLANG_NAMESPACE_BEGIN
 
-MemoryPool::MemoryPool(Arch arch) : arch_(arch) {
+MemoryPool::MemoryPool(Arch arch, Device *device)
+    : arch_(arch), device_(device) {
```

Comment:
It's a bit unclear to me if how arch_ and device_ should align in MemoryPool class. I'm definitely okay with this as an intermediate state but would love to see a comment here that describes desired end state. :D

Reply:
I understand your concern and I find it a bit unclear as well. The desired end state is to move the MemoryPool inside each Device, so that the memory allocated from each Device can be used as-is. I will add a comment as well.